libs/product/src/selectors/configurable-product/configurable-product.selectors.ts

```diff
@@ -52,41 +55,72 @@ const createConfigurableProductSelectors = (): DaffConfigurableProductMemoizedSe
 		selectProductEntities,
 		selectConfigurableProductAppliedAttributesEntitiesState,
 		(products, appliedAttributesEntities, props) => {
-			const product: DaffConfigurableProduct = products[props.id];
+			const product = <DaffConfigurableProduct>selectProduct.projector(products, { id: props.id });
 			if(product.type !== DaffProductTypeEnum.Configurable) {
 				return {};
 			}
-			const appliedAttributes: Dictionary<string> = selectConfigurableProductAppliedAttributes.projector(appliedAttributesEntities, { id: props.id });
-			const matchingVariants: DaffConfigurableProductVariant[] = selectMatchingConfigurableProductVariants.projector(products, appliedAttributesEntities, { id: props.id });
-			const unselectedAttributes = product.configurableAttributes
-				.filter(option => !appliedAttributes[option.code]);
+			const appliedAttributes: DaffConfigurableProductEntityAttribute[] = selectConfigurableProductAppliedAttributes.projector(appliedAttributesEntities, { id: props.id });
+			if(appliedAttributes.length === 0) return selectAllConfigurableProductAttributes.projector(products, { id: props.id });
 			
-			const undeterminedAttributes = {};
-			unselectedAttributes.forEach((attribute) => {
-				undeterminedAttributes[attribute.code] = [];
-				matchingVariants.forEach(variant => {
-					if(!isVariantAttributeMarkedAsUndetermined(undeterminedAttributes[attribute.code], variant.appliedAttributes[attribute.code])) {
-						undeterminedAttributes[attribute.code].push(variant.appliedAttributes[attribute.code]);
-					}
-				})
+			const selectableAttributes = initializeSelectableAttributes(product.configurableAttributes);
+			let matchingVariants: DaffConfigurableProductVariant[] = product.variants;
+			// Set which values of applied attribute codes should be set as selectable based on the order that they were selected
+			for(let i=0; i<=appliedAttributes.length; i++) {
+				matchingVariants = matchingVariants.filter(variant => isVariantAvailable(appliedAttributes.slice(0, i), variant));
+				if(i<appliedAttributes.length) {
+					matchingVariants.forEach(variant => {
+						if(!isVariantAttributeMarkedAsSelectable(selectableAttributes[appliedAttributes[i].code], variant.appliedAttributes[appliedAttributes[i].code])) {
+							selectableAttributes[appliedAttributes[i].code].push(variant.appliedAttributes[appliedAttributes[i].code]);
+						}
+					});
+				}
+			}
+			
+			// Set which values of the unapplied attribute codes should be set as selectable based on the matching variants of all
+			// applied attributes.
+			product.configurableAttributes.forEach(attribute => {
+				if(!selectableAttributes[attribute.code].length) {
+					matchingVariants.forEach(variant => {
+						if(!isVariantAttributeMarkedAsSelectable(selectableAttributes[attribute.code], variant.appliedAttributes[attribute.code])) {
+							selectableAttributes[attribute.code].push(variant.appliedAttributes[attribute.code]);
+						}
+					});
+				}
 			})
```

Comment:
This code is extremely difficult to read and understand. Mutating variables external to the loop scope makes it very difficult to reason about data flow and is generally considered to be an anti-pattern. You should endeavor to use pure functions and in-and-out data flow wherever possible, like so:
```ts
// Set which values of applied attribute codes should be set as selectable based on the order that they were selected
const matchedVariants = appliedAttributes.reduce((matchingVariants, appliedAttribute, i) => {
  const filteredVariants = matchingVariants.filter(variant => isVariantAvailable(appliedAttributes.slice(0, i), variant));

  selectableAttributes[appliedAttribute.code] = getSelectableAttributesFromVariants(selectableAttributes, filteredVariants, appliedAttribute.code);

  return filteredVariants
}, product.variants).filter(variant =>
  isVariantAvailable(appliedAttributes, variant)
);

// Set which values of the unapplied attribute codes should be set as selectable based on the matching variants of all
// applied attributes.
product.configurableAttributes.forEach(attribute => {
  if (!selectableAttributes[attribute.code].length) {
    selectableAttributes[attribute.code] = getSelectableAttributesFromVariants(selectableAttributes, matchedVariants, attribute.code);
  }
});

function getSelectableAttributesFromVariants(selectableAttributes: Dictionary<string[]>, variants: DaffConfigurableProductVariant[], code: string) {
  return variants.reduce((selectedAttributes, variant) =>
    isVariantAttributeMarkedAsSelectable(selectedAttributes, variant.appliedAttributes[code])
      ? selectedAttributes
      : [
        ...selectedAttributes,
        variant.appliedAttributes[code]
      ],
    selectableAttributes[code]
  )
}
```

Reply:
yea, I did try to do those things. What I was missing was this: 
```
const matchedVariants = appliedAttributes.reduce((matchingVariants, appliedAttribute, i) => {
  ...
}, product.variants).filter(...);
```

I stared at this part of my code for at least an hour just trying to solve this problem, but I couldn't figure out how to initialize the matchedVariants like that. The reduce function is pretty new to me, so thanks for showing me. changed